### DISCUSSION TITLE
使用debug模式编译时，如何去掉 -O0 编译选项？

### DISCUSSION THREAD

**Discussion by xumaofeng-jn** [xmake-io/xmake#4083] [Q&A] [answered]
目前发现项目从cmake切换到xmake后，debug版本编译速度明显下降，比较编译选项后，怀疑是 xmake 加的 -O0 导致的。
cmake 编译 debug 版本时，没有 -O0 编译选项，而 xmake 加上了 -O0 编译选项。
请问如何移除这种 xmake 内置的编译选项呢？target 只能 add_cxxflags，好像不能覆盖。

**Comment by waruqi** [comment]
去掉 `mode.debug` rule ，或者

```lua
if is_mode("debug") then
    set_optimize("fastest")
end
```

不应该是 `-O3`, `-O2` 更慢么。。`-O0` 禁用了优化，怎么会反而慢

**Comment by xumaofeng-jn** [comment]
@waruqi 
是编译速度更慢了，不是运行速度更慢。
加了优化选项后，编译器要对源文件执行优化，因此编译速度会更慢。
我期望的是set_optimize("none")后，不加 -O0选项，也不加其他任何 -Ox 优化选项，就和目前 cmake 做的一样。

**Reply by waruqi** [reply, marked as answer]
> -O0： 不做任何优化，这是默认的编译选项。
> -O0 Reduce compilation time and make debugging produce the expected results. This is the default.
https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html

-O0 原本就是不做任何优化 ，你可以看下 gcc 说明，即使不加，默认也是 -O0

你非要去掉，就 去掉 mode.debug rule 就行了。或者定义自己的 mode ，不用内置 debug rule

**Comment by xumaofeng-jn** [comment]
@waruqi 
你说的对，-O0 和 不加任何 -Ox 选项效果是一样的，前期怀疑 -O0 也会引入一些代码优化，影响编译速度。
后面比较发现编译结果磁盘占用比cmake占用大，发现xmake默认启用了ccache，禁用 ccache 后性能就差不多了。
set_policy("build.ccache", false)

ccache对于一个环境上有多个项目共用一个ccache是能加速编译性能的，但是我目前的编译环境仅有我一个人用，并且只有一个项目，意义不大，就暂时关闭了。
